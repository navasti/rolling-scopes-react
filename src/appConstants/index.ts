import { FormFields } from 'types';
import * as yup from 'yup';

export enum Limits {
  pokemon = 20,
  type = 10,
  move = 15,
}

export enum ErrorStatuses {
  success = '2',
  client = '4',
  server = '5',
}

export enum AvailableTabs {
  pokemons = 'pokemons',
  moves = 'moves',
  types = 'types',
}

export enum Fields {
  birthday = 'birthday',
  consent = 'consent',
  gender = 'gender',
  avatar = 'avatar',
  shiny = 'shiny',
  name = 'name',
  type = 'type',
}

export enum ErrorMessages {
  name = 'Name must be at least 2 characters long.',
  consent = 'Regulations consent is required.',
  birthday = 'Birthday is required.',
  type = 'Main type is required.',
  gender = 'Gender is required.',
}

export enum Types {
  fire = 'fire',
  grass = 'grass',
  poison = 'poison',
  flying = 'flying',
  ground = 'ground',
  psycho = 'psycho',
  water = 'water',
  ice = 'ice',
}

export const NONE = 'none';
export const MALE = 'male';
export const FEMALE = 'female';
export const INPUT_VALUE_KEY = 'input-value';
export const SUCCESS_MESSAGE = 'Pokemon added successfuly';
export const LINKS = ['home', 'about', 'form', 'invalid', 'details'] as const;
export const SEARCH_BAR_INSTRUCTIONS = [
  'Type and press enter to search for specific pokemon, type or move.',
  'Clear input and press enter to search for all pokemons, types and moves.',
] as const;

export const TABS = Object.values(AvailableTabs);
export const POKEMON_TYPES = Object.values(Types);

export const FORM_VALIDATION_SCHEMA = yup.object().shape({
  type: yup.string().required(ErrorMessages.type),
  name: yup.string().min(2, ErrorMessages.name).required(ErrorMessages.name),
  gender: yup.string().nullable().required(ErrorMessages.gender),
  avatar: yup
    .mixed()
    .test('fileSize', 'The file is too large', (value) =>
      !value?.length ? true : value?.[0]?.size <= 200000
    ),
  birthday: yup.string().required(ErrorMessages.birthday),
  consent: yup.bool().oneOf([true], ErrorMessages.consent),
  shiny: yup.bool(),
});

const BASE = 'https://pokeapi.co/api/v2';

export const API = {
  POKEMON_LIMIT: `?limit=${Limits.pokemon}`,
  TYPE_LIMIT: `?limit=${Limits.type}`,
  MOVE_LIMIT: `?limit=${Limits.move}`,
  POKEMON: `${BASE}/pokemon`,
  TYPE: `${BASE}/type`,
  MOVE: `${BASE}/move`,
  BASE,
};

export const FIELDS_VALIDATION_BY_NAME = {
  TEXT: [Fields.name, Fields.birthday, Fields.type],
  CHECK: [Fields.consent, Fields.gender],
};

export const DEFAULT_VALUES: FormFields = {
  consent: false,
  avatar: null,
  gender: MALE,
  shiny: false,
  birthday: '',
  name: '',
  type: '',
};
