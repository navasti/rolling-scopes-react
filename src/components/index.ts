export { Loader } from './Loader';
export { Modal } from './Modal';
