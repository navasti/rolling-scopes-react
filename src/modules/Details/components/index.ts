export { MoveCardDetails } from './MoveCardDetails';
export { TypeCardDetails } from './TypeCardDetails';
export { PokemonCardDetails } from './PokemonCardDetails';
