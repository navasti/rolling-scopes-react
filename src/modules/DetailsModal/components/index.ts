export { TypeModalContent } from './TypeModalContent';
export { MoveModalContent } from './MoveModalContent';
export { PokemonModalContent } from './PokemonModalContent';
