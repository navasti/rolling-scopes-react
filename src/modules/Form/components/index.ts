export { BirthdayField } from './BirthdateField';
export { ErrorMessage } from './ErrorMessage';
export { ConsentField } from './ConsentField';
export { AvatarField } from './AvatarField';
export { GenderField } from './GenderField';
export { ShinyField } from './ShinyField';
export { TypeField } from './TypeField';
export { NameField } from './NameField';
export { FormCard } from './FormCard';
