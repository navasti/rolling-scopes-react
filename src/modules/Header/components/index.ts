export { LocationBadge } from './LocationBadge';
export { Navigation } from './Navigation';
