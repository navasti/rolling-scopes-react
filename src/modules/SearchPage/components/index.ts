export { SearchBar } from './SearchBar';
export { Cards } from './Cards';
export { Card } from './Card';
export { Tabs } from './Tabs';
