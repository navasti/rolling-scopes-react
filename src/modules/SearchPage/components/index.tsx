export { SearchBar } from './SearchBar';
export { Card } from './Card';
