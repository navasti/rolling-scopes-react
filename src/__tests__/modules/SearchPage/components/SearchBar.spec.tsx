import { screen, render, fireEvent } from '@testing-library/react';
import { SearchBar } from 'modules/SearchPage/components';
import { SEARCH_BAR_INSTRUCTIONS } from 'appConstants';

const onSearchChange = jest.fn();
const onSearchKeyDown = jest.fn();

const getInput = (label: string): HTMLInputElement => screen.getByLabelText(label);

describe.skip('SearchBar', () => {
  afterEach(() => {
    onSearchChange.mockReset();
    onSearchKeyDown.mockReset();
  });
  it('search bar should display passed label, input value and match snapshot', () => {
    const { container } = render(
      <SearchBar
        inputValue="input-value"
        inputDisabled={false}
        label="search-bar-label"
        onChange={onSearchChange}
        onKeyDown={onSearchKeyDown}
      />
    );
    expect(getInput('search-bar-label')).toHaveValue('input-value');
    expect(getInput('search-bar-label')).not.toBeDisabled();
    expect(container).toMatchSnapshot();
  });
  it('search bar input should be disabled when isLoading prop equals true', () => {
    render(
      <SearchBar
        inputValue="input-value"
        inputDisabled={false}
        label="search-bar-label"
        onChange={onSearchChange}
        onKeyDown={onSearchKeyDown}
      />
    );
    expect(getInput('search-bar-label')).toBeDisabled();
  });
  it('search bar should invoke onChange method when input value changes', () => {
    render(
      <SearchBar
        inputValue="input-value"
        inputDisabled={false}
        label="search-bar-label"
        onChange={onSearchChange}
        onKeyDown={onSearchKeyDown}
      />
    );
    expect(onSearchChange).toBeCalledTimes(0);
    fireEvent.change(getInput('search-bar-label'), {
      target: { value: 'aa' },
    });
    expect(onSearchChange).toBeCalledTimes(1);
  });

  it('search bar should invoke onKeyDown method when any key is pressed', () => {
    render(
      <SearchBar
        inputValue="input-value"
        inputDisabled={false}
        label="search-bar-label"
        onChange={onSearchChange}
        onKeyDown={onSearchKeyDown}
      />
    );
    expect(onSearchKeyDown).toBeCalledTimes(0);
    fireEvent.keyDown(getInput('search-bar-label'));
    expect(onSearchKeyDown).toBeCalledTimes(1);
  });
  it('search bar input with icon should be rendered and visible', () => {
    render(
      <SearchBar
        inputValue="input-value"
        inputDisabled={false}
        label="search-bar-label"
        onChange={onSearchChange}
        onKeyDown={onSearchKeyDown}
      />
    );
    expect(screen.getByTestId('search-icon')).toBeVisible();
    expect(getInput('search-bar-label')).toBeVisible();
  });
  it('search bar instructions should be rendered and visible', () => {
    render(
      <SearchBar
        inputValue="input-value"
        inputDisabled={false}
        label="search-bar-label"
        onChange={onSearchChange}
        onKeyDown={onSearchKeyDown}
      />
    );
    SEARCH_BAR_INSTRUCTIONS.forEach((instruction) => {
      expect(screen.getByText(instruction)).toBeInTheDocument();
      expect(screen.getByText(instruction)).toBeVisible();
    });
  });
});
