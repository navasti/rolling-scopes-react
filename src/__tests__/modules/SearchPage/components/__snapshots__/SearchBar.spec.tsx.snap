// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBar search bar should display passed label, input value and match snapshot 1`] = `
<div>
  <div
    class="sc-bBrHrO jrHWRb"
  >
    <label
      class="sc-cxabCf cnuyDT"
      for="storage-input"
    >
      search-bar-label
    </label>
    <div
      class="sc-ivTmOn jWKPgk"
    >
      <svg
        data-testid="search-icon"
      >
        search.svg
      </svg>
      <input
        class="sc-llJcti cHKmYK"
        id="storage-input"
        placeholder="Type here"
        type="text"
        value="input-value"
      />
    </div>
  </div>
  <div
    class="sc-iIPllB fBkesa"
  >
    <span
      class="sc-gicCDI bkbDix"
    >
      Type and press enter to search for specific pokemon, type or move.
    </span>
    <span
      class="sc-gicCDI bkbDix"
    >
      Clear input and press enter to search for all pokemons, types and moves.
    </span>
  </div>
</div>
`;
